### PR TITLE
feat(mock): extend matchers

### PR DESCRIPTION
**Pull request by dsseng**
Add more matchers for mock functions

### CODE REVIEW

File: packages/mock/src/mockMatchers.ts [hunk 1]
```diff
@@ -51,6 +51,15 @@ const extension = {
     const nthCall = m.calls[n];
     return isEqual(nthCall.slice(0, args.length), args);
   },
+
+  toHaveReturned: (m: unknown, ...args: any[]): boolean => {
```
Comment (dsseng) [inline packages/mock/src/mockMatchers.ts]: Hmm, this logic seems to be the copy of `toHaveBeenCalled`, so probably this should be like this:
```suggestion
  toHaveReturned: (m: unknown, value: any): boolean => {
```
And we should search for `value` in `m.returns`, like in `toHaveBeenCalledWith`

File: packages/mock/test/mockMatchers.spec.ts [hunk 2]
```diff
@@ -76,4 +76,13 @@ describe('Mock function matchers', () => {
...
     expect(() => $expect(mockFn).toHaveBeenNthCalledWith(2, 3)).not.toThrow();
   });
+
+  it('toHaveReturned', () => {
+      expect(() => $expect({}).toHaveReturned()).toThrow(TypeError);
+
+      const mockFn = mock((a: any, b: any) => a + b);
+
+      mockFn(1, 2);
+      expect(() => $expect(mockFn).toHaveReturned()).not.toThrow();
```
Comment (dsseng) [inline packages/mock/test/mockMatchers.spec.ts]: This test should also test for conditions in which it __will__ throw

File: packages/mock/src/mockMatchers.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    n--;
+    if (n < 0 || n >= m.returns.length) {
+      return false;
+    }
+
+    const nthReturn = m.returns[n];
+    return isEqual(nthReturn, returnValue);
+  },
+
+  toHaveReturnedTimes: (m: unknown, times: number): boolean => {
```
Comment (dsseng) [inline packages/mock/src/mockMatchers.ts:95]: This doesn't make a lot of sense, it'll return the same as `toHaveBeenCalledTimes`.

File: packages/mock/src/mockMatchers.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    n--; // start with 1
+    if (n < 0 || n >= m.calls.length) {
+      return false;
+    }
+
+    const nthCall = m.calls[n];
+    return isEqual(nthCall.slice(0, args.length), args);
+  },
+
+  toHaveReturned: (m: unknown): boolean => {
```
Comment (dsseng) [inline packages/mock/src/mockMatchers.ts]: Doesn't make sense, clone of `toHaveBeenCalled`

File: packages/mock/src/mockMatchers.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  },
+
+  toHaveReturnedTimes: (m: unknown, times: number): boolean => {
+    if (!isMock(m)) {
+      throw new TypeError('expected value is not a mock function');
+    }
+
+    return isEqual(m.returns.length, times);
+  },
+};
```
Comment (dsseng) [inline packages/mock/src/mockMatchers.ts:119]: Maybe we should add smth like `toHaveReturnedTimes (m: unknown, times, value)` that'll check if this return value was returned n times?

File: packages/mock/src/mockMatchers.ts (hunk 3 above)
Comment (ElijahKotyluk) [reply packages/mock/src/mockMatchers.ts:95]: It is checking to see if the mock function has returned `x` times, it does not include thrown errors, throw causes this to fail.(See Jest) I just added a mock matcher to check if `v:any` has been returned `x:number` times: toHaveReturnedTimesWith

**Comment by ElijahKotyluk** [comment]
@sh7dm, Please review when you have the time to do so. 👍 
I will be able to fix any mistakes tonight.

File: packages/mock/src/mockMatchers.ts (hunk 3 above)
Comment (dsseng) [reply packages/mock/src/mockMatchers.ts:95]: let's keep it

**Comment by ElijahKotyluk** [comment]
@sh7dm Made minor changes, removed toHaveReturned mock matcher. If it checks out with you, I will merge.

**Review by ElijahKotyluk — COMMENTED** [review]
LGMT

File: packages/mock/src/mock.ts [hunk 6]
```diff
@@ -1,25 +1,45 @@
+export const mockSymbol = Symbol('isMock');
+
 // tslint:disable-next-line:interface-name
 export interface Mock extends Function {
   calls: any[][];
   returns: any[];
+  errors: any[];
```
Comment (dsseng) [inline packages/mock/src/mock.ts]: Errors aren't in scope of this. Could you please move them into a separate PR/commit to `master`?

File: packages/mock/test/mock.spec.ts [hunk 7]
```diff
@@ -32,4 +32,31 @@ describe('Mock function', () => {
...
+    expect(isMock(mockFn)).toBeTruthy();
+    expect(isMock(fn)).toBeFalsy();
+    expect(isMock(null)).toBeFalsy();
+    expect(isMock({})).toBeFalsy();
+    expect(isMock({
+      [mockSymbol]: true,
+    })).toBeFalsy();
+  });
+
+  it('should throw error', () => {
```
Comment (dsseng) [inline packages/mock/test/mock.spec.ts]: As I noted before, you'd better move it to separate pr/commit.

**Review by dsseng — COMMENTED** [review]
Looks good.
